parser_hevc.c

```diff
@@ -0,0 +1,387 @@
+// Tested with German DVB-T2 HD channels
+
+#include "parser_hevc.h"
+#include "bitstream.h"
+#include "config.h"
+
+#include <stdlib.h>
+#include <assert.h>
+
+
+cParserHEVC::cParserHEVC(int pID, cTSStream *stream, sPtsWrap *ptsWrap, bool observePtsWraps)
+ : cParser(pID, stream, ptsWrap, observePtsWraps)
+{
+  m_Height            = 0;
+  m_Width             = 0;
+  m_FpsScale          = 0;
+  m_PixelAspect.den   = 1;
+  m_PixelAspect.num   = 0;
+  memset(&m_streamData, 0, sizeof(m_streamData));
+  m_PesBufferInitialSize      = 240000;
+
+  m_IsVideo = true;
+  Reset();
+}
+
+cParserHEVC::~cParserHEVC()
+{
+}
+
+void cParserHEVC::Parse(sStreamPacket *pkt, sStreamPacket *pkt_side_data)
+{
+  if (m_PesBufferPtr < 10) // 2*startcode + header + trail bits
+    return;
+
+  int p = m_PesParserPtr;
+  uint32_t startcode = m_StartCode;
+  bool frameComplete = false;
+
+  while ((m_PesBufferPtr - p) > 3)
+  {
+    if ((startcode & 0x00ffffff) == 0x00000001)
+    {
+      if (Parse_HEVC(p, frameComplete) < 0)
+        break;
+    }
+    startcode = startcode << 8 | m_PesBuffer[p++];
+  }
+  m_PesParserPtr = p;
+  m_StartCode = startcode;
+
+  if (frameComplete)
+  {
+    if (!m_NeedSPS && m_FrameValid)
+    {
+      double PAR = (double)m_PixelAspect.num/(double)m_PixelAspect.den;
+      double DAR = (PAR * m_Width) / m_Height;
+      DEBUGLOG("HEVC SPS: PAR %i:%i", m_PixelAspect.num, m_PixelAspect.den);
+      DEBUGLOG("HEVC SPS: DAR %.2f", DAR);
+
+      int duration;
+      if (m_curDTS != DVD_NOPTS_VALUE && m_prevDTS != DVD_NOPTS_VALUE && m_curDTS > m_prevDTS)
+        duration = m_curDTS - m_prevDTS;
+      else
+        duration = m_Stream->Rescale(20000, 90000, DVD_TIME_BASE);
+
+      if (m_FpsScale == 0)
+        m_FpsScale = m_Stream->Rescale(duration, DVD_TIME_BASE, 90000);
+
+      bool streamChange = m_Stream->SetVideoInformation(m_FpsScale, DVD_TIME_BASE, m_Height, m_Width, DAR);
+
+      pkt->id       = m_pID;
+      pkt->size     = m_PesNextFramePtr;
+      pkt->data     = m_PesBuffer;
+      pkt->dts      = m_DTS;
+      pkt->pts      = m_PTS;
+      pkt->duration = duration;
+      pkt->streamChange = streamChange;
+
+    }
+    m_StartCode = 0xffffffff;
+    m_PesParserPtr = 0;
+    m_FoundFrame = false;
+    m_FrameValid = true;
+  }
+}
+
+void cParserHEVC::Reset()
+{
+  cParser::Reset();
+  m_StartCode = 0xffffffff;
+  m_NeedSPS = true;
+  m_NeedPPS = true;
+  memset(&m_streamData, 0, sizeof(m_streamData));
+  m_rbspBufferSize = 0;
+  if (m_rbspBuffer)
+  {
+    free(m_rbspBuffer);
+    m_rbspBuffer = NULL;
+  }
+}
+
+int cParserHEVC::Parse_HEVC(int buf_ptr, bool &complete)
+{
+  unsigned int len = m_PesBufferPtr - buf_ptr;
+  uint8_t *buf = m_PesBuffer + buf_ptr;
+  unsigned int p=0, NumBytesInNalUnit=0;
+
+  // find length of nal (including nal_unit_header)
+  uint32_t startcode = 0xffffffff;
+  while ((p+3) < len)
+  {
+    if ((startcode & 0x00ffffff) == 0x00000001)
```

Comment:
This was my idea for optimization: After the parsing, we know the position of the next start code, so the pointer could be set to this position immediately.
However, this will still require the accumulation of several TS packets until a complete NAL is found.